snap7/server.py

```diff
@@ -293,14 +294,75 @@ def clear_events(self) -> int:
 def mainloop(tcpport: int = 1102):
     server = snap7.server.Server()
     size = 100
-    DBdata = (snap7.types.wordlen_to_ctypes[snap7.types.S7WLByte] * size)()
-    PAdata = (snap7.types.wordlen_to_ctypes[snap7.types.S7WLByte] * size)()
-    TMdata = (snap7.types.wordlen_to_ctypes[snap7.types.S7WLByte] * size)()
-    CTdata = (snap7.types.wordlen_to_ctypes[snap7.types.S7WLByte] * size)()
+    DBdata = (snap7.types.wordlen_to_ctypes[snap7.types.WordLen.Byte.value] * size)()
+    PAdata = (snap7.types.wordlen_to_ctypes[snap7.types.WordLen.Byte.value] * size)()
+    TMdata = (snap7.types.wordlen_to_ctypes[snap7.types.WordLen.Byte.value] * size)()
+    CTdata = (snap7.types.wordlen_to_ctypes[snap7.types.WordLen.Byte.value] * size)()
     server.register_area(snap7.types.srvAreaDB, 1, DBdata)
     server.register_area(snap7.types.srvAreaPA, 1, PAdata)
     server.register_area(snap7.types.srvAreaTM, 1, TMdata)
     server.register_area(snap7.types.srvAreaCT, 1, CTdata)
+
+    ba = bytearray(1000)
```

Comment:
i think we should move this to a separate function, and make it optional to initialise these values.

Reply:
Agreed.
Also, how we should standarize the locations for each values? 
Example:
Booleans from byte 0-10?
Int from 10-20? 
and so on?

Reply:
i have no strong opinion about this.

Reply:
I would say, a simple structure, like you mentioned, would be nice. I like such rows to ten, easier to read and we are not restricted to tight space managament here.

Reply:
Ok, I will arrange it.
About the REAL, I had a lot of problems where I could not fix it.
I wrote in the server mainloop the real 3.123456e38 but when i read it in the client side i receive 3.1234567891e38 (the values are not real)

Reply:
This is the error that i'm getting it with real values.
```py
 def test_read_real(self):
        data = self.client.db_read(0, 37, 4 * 9)
        self.assertAlmostEqual(get_real(data, 0), -3.402823e38)
> AssertionError: -3.4028230607370965e+38 != -3.402823e+38 within 7 places (6.073709657548248e+30 difference)
```
I write the floating value directly to the bytearray like this
```py
ba[37:41] = struct.pack(">f", -3.402823e38)
```

Reply:
It seems like the `get_real` is the problem, because if i compare the bytearray before using the utils function both are the same
 ```py
    def test_read_real(self):
        data = self.client.db_read(0, 37, 4 * 9)
        print(data[0:4])
        self.assertEqual(data[0:4], b'\xff\x7f\xff\xfd')
>       self.assertAlmostEqual(get_real(data, 0), -3.402823e38)
E       AssertionError: -3.4028230607370965e+38 != -3.402823e+38 within 7 places (6.073709657548248e+30 difference)

test/test_mainloop.py:103: AssertionError
------------------------------------------------- Captured stdout call -------------------------------------------------
bytearray(b'\xff\x7f\xff\xfd')
```

Reply:
It seems that for some of the numbers provided by @swamper123 (ex: -3.402823e38) the presition is lost in single presition and we get -3.4028229e+38
![image](https://user-images.githubusercontent.com/43883304/111547389-d75c7f80-8757-11eb-9613-362c985bc944.png)
In the test before the difference between the real value, and the read is 0.00001% (nothing) 
How should i proceed?

Reply:
Different sources, different numbers, but all from the forum on the siemens homepage (*grumble grumble*). 
Since REAL is 32bit long, single presition seems to be the right answer.

Reply:
Yes single presition is correct, i think it is a problem of presition for that big number, I added an assert almost equal with a delta of 0.0001% 

And for the string, I didn't understand what you said about the first two bytes, but it seems that the first is the total length (n+2) and the second one is the actual length (n)
![image](https://user-images.githubusercontent.com/43883304/111610638-57b3cc80-87ba-11eb-827c-6c3598964891.png)

Reply:
STRING[254] is the same as STRING.
n is the number of chars in the strings.

So STRING[254] is 254+2 Bytes big -> max. 256 Bytes

Like in the picture, Byte0 contains max length of a string and Byte1 the actural length of the string. In Byte3 to Byte255 contain the actual ASCII character.